README.md
MCK: The ST Mint Construction Kit
=================================

A proof of concept to build an Atari ST Freemint distribution using Ansible as a build tool.
Freemint is an Unix-like kernel for the Atari ST, see https://freemint.github.io/ for details

## Ready to use disk images for your Atari
See http://www.subsole.org/st_mint

## Usage

Having ansible and a POSIX make installed, call

```
make
```

To copy all the stuff in the first partition of a TOS/DOS disk image named 1GB.img, call

```
make install
```

To copy all the stuff in a partition mounted on /stmint, call
```
make cardinstall
```

## Roadmap / Status
see https://github.com/EmmanuelKasper/mck/wiki/Roadmap




csed/changelog.txt
2003-08-15
- *, \+, \?, \{... treated as \*, +, ?, { when not following a 
  repeatable expression.
- implemented [.a.], [=a=] and [:alpha:] for the POSIX locale only; 
  numerous improvements in the bracket expression compiler.
- removed the bug-o-matic files, put them in another project.
- \< and \> no longer look outside allocated memory.
- reorganised the todo list.
- updated the README.
- added a copy of the GPL; added GPL ritual invocations on top of files.

2003-08-14
- made buffers, compiled commands and labels dynamically allocated. This
  was a rather big change, but I really wanted to remove size limitations. 
  Speed does not seem to be affected.
- changed the way script lines are read: only one cmdline() function now,
  allowing cleaner script line handling.
- location of syntax errors reported
- command encoding changed: number of line addresses no longer limited; 
  N flag to s/// no longer limited to 4096.
- selected() entirely rewriten: now at last I understand how it works.
- PLUS address feature removed
- rewrote getline() to no longer use fgets()
- unlimited number of appends (from a\ and r commands)
- reworked test mechanisms: adding a new test file no longer re-tests all.
- addr.t - added address range tests
- made { ... } identical to a branch (previously skipping a {...} group
  would reinitialize the 'inrange' state of skipped range addresses)
- run-time error report includes script location when relevant

2003-08-13
- started breaking everything to add dynamic memory allocation...

2003-03-27
- error messages now moved directly where they're used in the code.
- better usage()
- removed last remaining /*hh NN */ comments, that did not mean anything
  any more after the code was indented and further modified by me.

2003-03-26
- test/8bit.sh - now uses the same convention of *.tmp and tmp.* for temp 
  files
- rewrote getline() with fgets() instead of gets: sed is now less likely
  to dump core when reading a too long input line.
- misc.t - more misc. tests
- gratuitiously reworked rhscomp()
- gettext split in two different functions (one for labels and filenames,
  one for the text argument of a\, c\, and i\ command) - the text after a\
  now corectly keeps its initial space.
- fixquote() reworked to have a more intuitive interface
- test.sed altered so that text out of entries must begin with '#'.
- *.t - modified to comply to the new test.sed syntax
- fixed //\2/ again (the previous test was badly processed by test.sed)
- misc.t - added test for i\ and for branching commands b and t

2003-03-25
- s//\1/ now allowed. The check is still made at compile time in cases
  like s/\(.\)/\1/, but the the RE is empty the test is made instead
  at runtime. 
- test/misc.t - test for dynamic // handling; test for s//\1/

2003-03-24
- guard.c: runs a program and kill it after a few seconds (to guard against
  infinite loops when testing)
- played with a.sh, the shell running the random tests

2003-03-23
- temporarily removed numbered subst from the grammar, as they 
  did raise too many times the same GNU sed bug
- a.sh - tried to find a way to determine if the sed command
  runs correctly, or fails, or crashes, or loops forever (and if so, kill 
  it after 5 seconds), but this doesn't work yet.

2003-03-22
- added an awk program that generates random sed tests.
- nested brackets like \(a\(b\)c\2\) now allowed.
- updated match.txt to cover nested brackets like \(a\(b\)c\2\)

2003-03-21
- \{0,\} now allowed as synonym to *
- fixed the Makefile
- \)* is now CMKET instead of CKET|STAR
- sedexec.c - removed macro GETNUM; tried some silly micro-optimisations
- sedexec.c :: substitute() - simplified code; took into account the fact 
  that s///3g is now forbidden; null matches now correctly handled: 
  s/b*/x/g gives xaxcx on abc
- sedcomp.c - removed place() for faster replacements
- sedcomp.c - better error message for s///2g; also disallowed s///2p3
- sedexec.c - fixed t command (used to only consider the very latest subst)
  -> dc.sed runs correctly now
- README - corrected thanks and authors sections
- *.c - moved record of changes as comments to the changelog
- *.[ch] - edited comments so that line are shorter than 80 chars
- formatted the fearsome cmdmask[] table and protected it from indent

2003-03-19
- allowed \{0,n\}
- added \? as a shortcut for \{0,1\}
- implemented repeated grouping \(...\)\{...\} and started some testing
- DEBUG option: -d option and #d as first comment
- removed illegal syntax /a/;/b/ that was redundant with /a/,/b/
- command lines are rejected if too long when read from files
- rewrote compile(): solved s///10;; bug and disallow s///10g or s///g10
- added #n in some scripts of the GNU test suite for convenience
- commented out madding test
- s/\(\)//g and s/\</x/g no longer loop forever
- \1* correctly handled when \1 is null

2003-03-17
- added a kind of README
- sedexec.c - removed silly macro IFEQ
- proper handling of multiple '!'s: /foo/!!!!d is the same as /foo/!d
- # made only valid after a command
- unified the way \{...\} and * are handled: the code is now more complex 
  when compiling, but smaller when executing (perhaps marginally slower)
  also started providing support for repeated grouping.
- added a test suite derived from that of GNU sed 4.0.5
- test.sed - new version with a syntax adapted so that it doesn't clash
  with the GNU test suite
  
2003-03-13
- revised RE and addresses encoding, separating them using different
  symbol names.
- bug solved: a*\1 was wrong when \1 was null
- // now refers to the last RE used (dynamically) instead of the
  last RE compiled.
- Makefile - adapted to work on both cygwin and Unix

2003-03-12
- used stdlib function instead of private smemcmp
- used unsigned chars everywhere internally
- numbers in \{n,m\} can now grow up to 32767 (previously <= 255)
- lastrhs misfeature removed 
  (s/foo/bar/;s/b/%/ used to be a shortcut for s/foo/bar/;s/b/bar/)
- test.sed - easier editing of test files


csed/8bit.sh
#
# test handling of 8bit chars and the l command
#

echo $sed '-f tmp.sed tmp.in > tmp.result 2>&1
if ! cmp -s tmp.out tmp.result
then
  echo ---- test failed when running sed script:
  escape -c% tmp.sed
  echo -- on following data:
  escape -c% tmp.in
  echo -- expected:
  escape -c% tmp.out
  echo -- but got:
  escape -c% tmp.result
  echo --
fi' > tmp.try

echo '\1\2\3\4\5\6\7
\10\11\12\13\14\15\16\17
\20\21\22\23\24\25\26\27
\30\31\32\33\34\35\36\37
 \\$$
\330\331\332\333\334\335\336\337' | escape -u > tmp.in


echo '\001\002\003\004\005\006\a$
\b\t$
\v\f\r\016\017$
\020\021\022\023\024\025\026\027$
\030\031\032\033\034\035\036\037$
 \\$$$
\330\331\332\333\334\335\336\337$' > tmp.out

echo 'l;d' > tmp.sed
. tmp.try

echo 'foo' > tmp.in
echo 'foo\nfoo$' > tmp.out
echo 'h;G;l;d' > tmp.sed
. tmp.try 


echo '%x01%x02%x03%x04abcdefbcdef%x7e%x7f%x80%x81%x82
%x83%x84%x85%x86%x87%x88%b%x89%x8a%x8b%x8c%x8d%x8e%x8f
%xfe%xff' | escape -uc% > tmp.in

echo 's/%x01%x02/foo/p
s/bcde/%a%b%f%r%t%v/g
y/%b%f/AZ/
s/[%x84-%x8c]\{1,\}/x/g
s/^..$/&&&/' | escape -uc% > tmp.sed

echo 'foo%x03%x04abcdefbcdef%x7e%x7f%x80%x81%x82
foo%x03%x04a%aAZ%r%t%vf%aAZ%r%t%vf%x7e%x7f%x80%x81%x82
%x83xAx%x8d%x8e%x8f
%xfe%xff%xfe%xff%xfe%xff' | escape -uc% > tmp.out

. tmp.try

# \x80 not handled by hhsed
echo 's/foo/%x800/' | escape -uc% > tmp.sed
echo 'foo' > tmp.in
echo '%x800' | escape -uc% > tmp.out
. tmp.try



